### ISSUE TITLE
Support for images from a browser

### ISSUE THREAD

**Issue by ttmx** [sentriz/cliphist#28]
Copying an image from the web and running
```
$ wl-paste -l
image/png
text/html
text/x-moz-url
```
And after re-copying it from cliphist
```
$ wl-paste -l
text/plain
text/plain;charset=utf-8
TEXT
STRING
UTF8_STRING
```

It only seems to preserve the `text/html` part of the clipboard.

It would be very nice if cliphist managed to keep all the properties, meaning we could copy old images from the web from the history.

Thank you for the nice software so far :)

**Comment by sentriz** [comment]
thanks!

it works for me and firefox at least

do you have xdg-utils installed?

```shell
$ wl-paste -l
text/html
text/_moz_htmlinfo
text/_moz_htmlcontext
image/png
image/jpeg
image/bmp
image/x-bmp
image/x-MS-bmp
image/x-icon
image/x-ico
image/x-win-bitmap
image/vnd.microsoft.icon
application/ico
image/ico
image/icon
text/ico
image/tiff
SAVE_TARGETS
```

**Comment by sentriz** [comment]
oh wait. sorry i missed the bit about recopying. that's weird since when i copy an image the actual bytes are copied. and cliphist shows "binary data image/png". then can recopy and paste fine

**Comment by ttmx** [comment]
Try copying it from a chromium based browser, not just from `wl-copy < img.png` or an equivalent. Is the behaviour still the same?

21 Oct 2022, 19:10 by ***@***.***:

**Comment by sentriz** [comment]
ah I see what you mean. what happens when you spawn a second wl-paste watch command? like

eg `wl-paste --type image --watch cliphist store`

then in your clipboard history you should see both the html and raw image data

**Comment by 7ff04da4aa6ac37197e249299066ac30** [comment]
Not the original poster, but spawning a second wl-paste watch command seems to work as expected, i.e. generating raw image data and html. Tested on both firefox and chromium, on Arch Linux. Considering I'm using sway, is it safe to assume that having, simultaneously, the lines `exec_always wl-paste --watch cliphist store` and `exec_always wl-paste --type image --watch cliphist store` in my config file will not generate any conflicts?

Edit: so far (some hours after), everything has been working as expected :)

**Comment by sentriz** [comment]
@7ff04da4aa6ac37197e249299066ac30 I think it should be fine. there's a lock in place on the db so only one can write at a time. also there is some deduplication logic

btw why do you want to exec_always? (as opposed to exec) I think with the always version you'll get a new wl-paste process every time you reload sway

**Comment by 7ff04da4aa6ac37197e249299066ac30** [comment]
@sentriz as far as I'm aware, it ensures that wl-paste is run every time I reload sway, which is important for the time being, as I've been messing with my configuration and I constantly reload it to test my modifications.

I have used `exec` previously, but I found that this killed the wl-paste process whenever I reloaded sway. At least, that was the case when using a waybar module that depends on both wl-clipboard and cliphist. Using `exec_always` was the measure that fixed this issue.
